Domain decomposition merge

liangjg opened the pull request:
This PR adds domain decomposition(dd) to OpenMC. To support this a number of changes have been made throughout the codebase - mostly minor, but some major.
- Adds basic domain decomposition classes and functions
- Adds on-the-fly memory loading of tally results data
- Changes from a single particle data structure used for tracking to an array of particles
- Changes source and fission bank arrays to re-sizable vectors
- Changes the random number scheme between batches - streams of parent neutrons are continued by the children.
- Updates the documentation for usage of domain decomposition, and otf-tally

Notes 1: The merged version will not change calculating results if dd is not activated. The calculations using dd with different parallel processes would get identical results, but they are different from non-dd runs because of the new random number scheme.

Note 2: Several features have been implemented but finally removed for this PR, such as another implementation of distributed material (input in material file), on-the-fly allocation of material data (and output material), and a unit testing framework. They can be found in the commits history.

wbinventor commented:
:+1 Awesome to see this PR @liangjg!!

siegela67 commented:
ditto. this is great.

On Jan 15, 2016, at 10:44 PM, Will Boyd <notifications@github.com> wrote:

> :+1 Awesome to see this PR @liangjg!!
> 
> —
> Reply to this email directly or view it on GitHub.
>

smharper commented:
Thanks for getting this going @liangjg!  I'll try to give you some comments today.

smharper commented on .gitignore:
```diff
@@ -19,6 +19,9 @@ examples/python/**/*.xml
 
+# vim backups
+*.swp
+
```
Remove these three lines.  They are redundant due to recent pull requests.

smharper commented on CMakeLists.txt:
```diff
@@ -248,7 +248,7 @@ set(CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
 
 set(program "openmc")
-file(GLOB source src/*.F90 src/xml/openmc_fox.F90)
+file(GLOB source src/*.F90 src/xml/openmc_fox.F90 ${test_source})
```
Reverse this change.  It was probably part of the unit tests.

smharper commented on src/source.F90:
```diff
@@ -12,10 +13,14 @@ module source
+  use random_lcg,       only: prn, set_particle_seed, prn_set_stream, prn_seed
   use state_point,      only: read_source_bank, write_source_bank
   use string,           only: to_str
+  use random_lcg_header,  only: STREAM_TRACKING, STREAM_SOURCE
```
Remove this `use`

smharper commented on src/source.F90:
```diff
@@ -12,10 +13,14 @@ module source
 
+#ifdef HDF5
+  use hdf5_interface
+#endif
```
Remove this `#ifdef` blcok

smharper commented on src/source.F90:
```diff
@@ -45,6 +50,11 @@ subroutine initialize_source()
 
+      if (dd_run) then
+        call fatal_error("Reading source from binary file not " // &
+                         "implementedfor domain decomposition.")
```
Add a space between "implemented" and "for"

smharper commented on src/state_point.F90:
```diff
@@ -60,15 +62,27 @@ subroutine write_state_point()
+    end if
+
+    ! Append appropriate extension
+
```
Remove this extra line after the comment

smharper commented on src/tally_header.F90:
```diff
@@ -124,6 +148,174 @@ module tally_header
+
+      this % n_triggers = 0
+
+    end subroutine tallyobject_clear
```
Remove `tallyfilter_clear` and `tallyobject_clear`

smharper commented on src/tally_header.F90:
```diff
@@ -55,8 +56,13 @@ module tally_header
   type TallyFilter
     integer :: type = NONE
     integer :: n_bins = 0
+    integer :: offset = 0 ! Only used for distribcell filters
```
Remove `offset`

smharper replied:
This line still needs to be removed

smharper commented on src/state_point.F90:
```diff
@@ -461,6 +694,11 @@ subroutine write_source_point()
     character(MAX_FILE_LEN) :: filename
 
+    if (dd_run) then
+      if (master) call warning('Source bank writing not implemented for DD runs.')
```
Make this a `fatal_error` and expand "DD" to "domain decomposed"

smharper replied:
I'm a little confused.  This warning suggest that writing the source is not implemented but then why is there all this DD specific code below?

smharper commented on src/state_point.F90:
```diff
@@ -487,7 +732,14 @@ subroutine write_source_point()
       else
         filename = trim(path_output) // 'statepoint.' // &
-             zero_padded(current_batch, count_digits(n_max_batches))
+             & zero_padded(current_batch, count_digits(n_batches))
```
Reverse the changes to this line

smharper commented on src/state_point.F90:
```diff
@@ -476,6 +714,13 @@ subroutine write_source_point()
+
+        if (dd_run) then
+          filename = trim(filename) // '.domain_' // &
+               & zero_padded(domain_decomp % meshbin, &
```
Remove the leading `&`

smharper commented on src/state_point.F90:
```diff
@@ -487,7 +732,14 @@ subroutine write_source_point()
+
+        if (dd_run) then
+          filename = trim(filename) // '.domain_' // &
+               & zero_padded(domain_decomp % meshbin, &
```
Remove the leading `&`

smharper commented on src/state_point.F90:
```diff
@@ -502,6 +754,12 @@ subroutine write_source_point()
       filename = trim(path_output) // 'source' // '.h5'
+      if (dd_run) then
+        filename = trim(filename) // '.domain_' // &
+             & zero_padded(domain_decomp % meshbin, &
```
Remove the leading `&`

smharper commented on src/state_point.F90:
```diff
@@ -785,6 +1055,154 @@ subroutine load_state_point()
+      end do MOMENT_LOOP
+
+    end do TALLY_METADATA
+
```
Remove all of these changes to `load_state_point`

smharper commented on src/state_point.F90:
```diff
@@ -1026,4 +1444,95 @@ subroutine read_source_bank(group_id)
+!===============================================================================
+
+  subroutine heapsort_results(t)
+
```
Remove this blank line

smharper commented on src/state_point.F90:
```diff
@@ -1026,4 +1444,95 @@ subroutine read_source_bank(group_id)
+!===============================================================================
+
+  subroutine siftdown_results(t, start, bottom)
+
```
Remove this blank line

smharper commented on src/tally.F90:
```diff
@@ -102,7 +102,6 @@ subroutine score_general(p, t, start_index, filter_index, i_nuclide, &
           score = flux
         end if
 
-
```
Add this blank line back in

smharper commented on src/tracking.F90:
```diff
@@ -187,6 +292,7 @@ subroutine transport(p)
             p % coord(j + 1) % uvw = p % coord(j) % uvw
           end if
         end do
+
```
Remove this blank line

smharper commented on src/tracking.F90:
```diff
@@ -114,6 +217,8 @@ subroutine transport(p)
              distance * material_xs % nu_fission
       end if
 
+
```
Remove this blank line

smharper commented on src/utils/test_template.py:
```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env python2
```
This file is part of the unit test system.  Remove it

smharper commented on src/utils/dumpmat.py:
```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python2
```
I think this file goes with the old distribmat system.  Remove it

smharper commented on src/utils/distribcell.py:
```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python2
```
This file is really out of date.  You could try to update it to the Python API and make it more usable, but I would recommend just deleting it.

smharper commented:
"test_domain_decomp_otf_tallies" is a really long name.  I would recommend changing it to "test_dd_otf_tallies".  Similarly, you can shorten "test_domain_decomp" to just "test_dd"

smharper commented on src/dd_init.F90:
```diff
@@ -0,0 +1,326 @@
+  use search,     only: binary_search
+  use string,     only: to_str
+
+#ifdef MP
```
Typo `#ifdef MP` -> `#ifdef MPI`.  This might fix your Travis errors

liangjg commented on src/dd_init.F90:
```diff
@@ -0,0 +1,326 @@
+
+#ifdef _OPENMP
+    call fatal_error("Domain decomposition not implemented in " // &
+         "conjunction with OpenMP.")
```
Thank you, @smharper . I found Travis tests failed because the two domain decomposition tests give errors when the code is compiled with openMP. So what's your suggestion to exclude these two tests for omp-compiling modes? Thanks.

smharper replied:
Ideally we would modify the code so it will work with OpenMP.  Do you know why it would not work with OpenMP?

paulromano replied:
In order to be considered for merging into the develop branch, the changes will need to work with OpenMP.

liangjg replied:
@smharper @paulromano  I tried to figure out why dd doesn't work with multi-threads. Several locations should account for threads parallel. One of those is the otf-tally, a key part of dd to distribute memory of tally data. If multi-threads transport different particles simultaneously,  the on-the-fly allocation of tally is possible to cause competition. So I made the allocation a critical section. 

Besides, I found maximum number of fission_bank size for thread can be reached in some cases, resulting in irreproducible results using different number of threads. I modified the size from "3*work/n_threads" to "5*work/n_threads".

smharper commented:
Please move the new examples into the "examples/xml" directory

liangjg commented:
The Travis tests failed bacause of "compiler errors" in 'omp-hdf5-debug' mode. But it passed on nsecluster. Could you tell me how to see the error messages on Travis? Or you may know the reason, right? Thanks for your help. @smharper @paulromano

paulromano commented:
@liangjg Try making the following change:
```diff
diff --git a/src/global.F90 b/src/global.F90
index 4ae1c59..b74d95d 100644
--- a/src/global.F90
+++ b/src/global.F90
@@ -23,7 +23,6 @@ module global
 #endif
 
   implicit none
-  save
 
   ! ============================================================================
   ! GEOMETRY-RELATED VARIABLES
```

liangjg commented:
Travis passed finally. Thanks for the help. @smharper @paulromano . Please consider remaining reviewing.

smharper commented:
@paulromano, can you elaborate on that module `save` attribute?  Why did we previously have it and why did it cause this failure?

smharper commented on src/state_point.F90:
```diff
@@ -269,6 +310,7 @@ subroutine write_state_point()
               call write_dataset(filter_group, "type", "delayedgroup")
             end select
 
+            call write_dataset(filter_group, "offset", tally%filters(j)%offset)
```
Remove this line

smharper commented on src/state_point.F90:
```diff
@@ -686,16 +928,28 @@ end subroutine write_tally_results_nr
+    integer, allocatable       :: temp_array(:)
+    integer                    :: n_order      ! loop index for moment orders
+    integer                    :: nm_order     ! loop index for Ynm moment orders
+    character(8)               :: moment_name  ! name of moment (e.g, P3, Y-1,1)
```
Remove all of these added variables

smharper commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1104 @@
+    integer :: to_rank        ! global rank of the processor to send to
+    integer :: alloc_err      ! allocation error code
+    integer :: n_request      ! number of communication requests
+    integer :: new_source     ! size of the source_bank after distribution
```
Make new_source an `integer(8)`

smharper commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1104 @@
+         call extend_array(source_bank, size_source_bank, int(new_source, 8), &
+             alloc_err)
+        call resize_array(dd % particle_buffer, dd % size_particle_buffer, &
+             int(new_source, 8), alloc_err)
```
Remove these `int` casts of `new_source`

smharper replied:
I don't like the `resize_array` routine.  It's not clear from the name that it can only make arrays larger and that it erases the contents.  I think it's also strange that it requires and then modifies the `current_size` argument.  Please replace all occurrences of `resize_array` with `deallocate` and `allocate` statements, and delete its definition in resize_arr.F90

liangjg replied:
@smharper , Thanks for your comments, especially the `VectorInt` suggestion, it is great. Now as for the `resize_array`, what if renaming it, for example to `extend_clear_array`? As this routine is an encapsulated interface and it is called at several places with different instances. Replacing it with `deallocate` and `allocate` statements may decrease neatness?  Besides,  `resize_array` seems coordinated with `extend_array`. If one is unfolded, the other one should be done as well?

smharper commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1104 @@
+    integer :: alloc_err      ! allocation error code
+    integer :: n_request      ! number of communication requests
+    integer :: new_source     ! size of the source_bank after distribution
+    integer, allocatable    :: request(:) ! communication requests
```
Replace this request with a `VectorInt` from stl_vector.F90

smharper commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1104 @@
+      n_request = n_request + 1
+#ifdef MPI
+      call MPI_IRECV(source_bank(n_source_sites), 1, MPI_BANK, MPI_ANY_SOURCE, &
+           MPI_ANY_TAG, MPI_COMM_WORLD, request(n_request), mpi_err)
```
When you switch `request` to a `VectorInt`, you'll need to replace this `request(n_request)` with some new variable, maybe `integer current_request` then use `call request % push_back(current_request)`

smharper commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1104 @@
+
+    ! Wait for all sends/recvs to complete
+#ifdef MPI
+    call MPI_WAITALL(n_request, request, MPI_STATUSES_IGNORE, mpi_err)
```
With `VectorInt` I guess you'll have to replace this with
```f90
call MPI_WAITALL(request % size(), request % data(1:request % size()), MPI_STATUSES_IGNORE, mpi_err)
```

smharper replied:
Or maybe just
```f90
call MPI_WAITALL(request % size(), request % data, MPI_STATUSES_IGNORE, mpi_err)
```

smharper commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1104 @@
+    integer :: second_neighbor_bin
+    integer :: to_rank        ! global rank of the processor to send to
+    integer :: from_rank      ! global rank of the processor to recv from
+    integer :: request(10000)   ! communication request for send/recving particles TODO this should be allocated to an appropriate max size depending on the number of procs per domain
```
The reccommended changes to `request` applies here as well

smharper commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1104 @@
+    integer :: neighbor_bin   ! domain meshbin of a neighbor to to_bin
+    integer :: index_scatt    ! index in the 'global' scatter array
+    integer :: domain_start
+    integer :: request(10000)   ! communication request for send/recving particles TODO this should be allocated to an appropriate max size depending on the number of procs per domain
```
Ditto `VectorInt`

smharper commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1104 @@
+    integer :: to_rank        ! global rank of the processor to send to
+    integer :: from_rank      ! global rank of the processor to recv from
+    integer :: index_pbuffer  ! index in the particle buffer
+    integer :: request(10000) ! communication request for send/recving particles TODO this should be allocated to an appropriate max size depending on the number of procs per domain
```
Ditto `VectorInt`

smharper commented on src/tracking.F90:
```diff
@@ -82,23 +116,92 @@ subroutine transport(p)
-      if (p % material /= p % last_material) call calculate_xs(p)
+      ! determine whether the particle is now transferred to right domain
+      if (dd_run .and. &
+           & p % stored_distance > ZERO) then
```
Remove the line break

smharper commented on src/tracking.F90:
```diff
@@ -82,23 +116,92 @@ subroutine transport(p)
+             distance, d_dd_mesh, meshbin=domain_decomp % meshbin)
+        p % coord(1) % xyz = xyz_temp
+        p % fly_dd_distance = p % fly_dd_distance + d_dd_mesh
+        if ( p % stored_distance - p % fly_dd_distance > FP_COINCIDENT) then
```
Remove the space after the `if (`

smharper commented on src/tracking.F90:
```diff
@@ -82,23 +116,92 @@ subroutine transport(p)
+
+      ! Check for domain boundary crossing
+      if (dd_run .and. &
+           d_dd_mesh <= distance) then
```
Remove the line break

smharper commented:
@liangjg why did you remove the example?

smharper commented on src/random_lcg_header.F90:
```diff
@@ -0,0 +1,12 @@
+  integer, parameter :: STREAM_SOURCE = 3
+
+end module random_lcg_header
+
```
Remove the unnecessary line at the end of the file

paulromano commented:
@smharper Regarding the module `save` attribute- my understanding is that prior to the Fortran 2003 standard, there was no guarantee that module variables would remain defined if they went out of scope. Thus, it was highly advisable to put a `save` statement in a module to tell the compiler that the module variables should remain defined and unchanged even if they go out of scope in a particular scoping unit. However, in F2003 and F2008, the standard guarantees that module variables implicitly have the `save` attribute, so it is superfluous to specify it. 

Now, as to why I suggested it be removed, I ran across a weird bug with gfortran 4.6 when working on #557 where it complained about a variable in the global module. I found that by removing the existing `save` statement, the error went away. I didn't see this problem with a newer version of gfortran.

paulromano commented on src/input_xml.F90:
```diff
@@ -5134,6 +5268,11 @@ subroutine expand_natural_element(name, xs, density, list_names, &
+! libary for addition to materials with depletion on
+!===============================================================================
+
+!===============================================================================
```
Remove this

paulromano commented on src/input_xml.F90:
```diff
@@ -1482,6 +1607,7 @@ subroutine read_geometry_xml()
       deallocate(coeffs)
 
       ! Boundary conditions
+      s % bc = BC_TRANSMIT
```
Why was this needed?

liangjg replied:
This line `s % bc = BC_TRANSMIT`  was added by Nick in a commit "Set default surface boundary condition to TRANSMIT". I think it is redundant too and will revert it.

paulromano commented on src/input_xml.F90:
```diff
@@ -699,6 +704,126 @@ subroutine read_settings_xml()
+
+        call get_node_array(node_dd, "nodemap", domain_decomp % domain_load_dist)
+
+        if (any(domain_decomp % domain_load_dist < 0.0_8 )) then
```
Use `ZERO` instead of `0.0_8`

paulromano commented:
You need to update the schema for input files in `src/relaxng/`. First you should update the .rnc files and then automatically generate the rng using trang, e.g.
```sh
trang settings.rnc settings.rng
```
If you don't have it installed, you can run `sudo apt-get install trang`. Let me know if you need help.

smharper commented:
Damn, I've been manually updating the rng files all this time.  Good to know

I'm going to preempt @paulromano and say that you also need to update the PythonAPI.  But please do not do that yet.  I am worried that adding PythonAPI diffs will make the diffs so long that GitHub will stop showing them.

liangjg commented:
I removed the examples in `examples\distribution` because they are for old distribmats and not for domain decomposition. @smharper

paulromano commented on src/geometry.F90:
```diff
@@ -366,6 +371,54 @@ recursive subroutine find_cell(p, found, search_cells)
+! CALIBRATE_COORD recalculates lower level coordinates based on level data, to
+! make sure coord() consistent with each other
+!===============================================================================
+subroutine calibrate_coord(p)
```
This lin should be intended two spaces

paulromano commented on src/geometry.F90:
```diff
@@ -366,6 +371,54 @@ recursive subroutine find_cell(p, found, search_cells)
+      end if
+    end do
+
+end subroutine calibrate_coord
```
This line should also be indented two spaces

paulromano commented on src/geometry.F90:
```diff
@@ -366,6 +371,54 @@ recursive subroutine find_cell(p, found, search_cells)
+
+    if(p % n_coord <= 1) then
+      return
+    end if
```
Replace this block with
```Fortran
if (p % n_coord <= 1) return
```

paulromano commented on src/dd_header.F90:
```diff
@@ -0,0 +1,109 @@
+  private
+  public :: deallocate_dd
+
+  type, public :: dd_type
```
Use CamelCase for derived type names and no underscores. See the [style guide](http://openmc.readthedocs.org/en/latest/devguide/styleguide.html#derived-types-and-classes)

paulromano commented on src/geometry.F90:
```diff
@@ -203,6 +207,7 @@ recursive subroutine find_cell(p, found, search_cells)
     type(Cell),     pointer :: c    ! pointer to cell
     class(Lattice), pointer :: lat  ! pointer to lattice
     type(Universe), pointer :: univ ! universe to search in
+    type(Material), pointer :: mat  ! pointer to material
```
Remove this

paulromano commented on src/geometry.F90:
```diff
@@ -5,6 +5,10 @@ module geometry
   use geometry_header,        only: Cell, Universe, Lattice, &
                                     &RectLattice, HexLattice
   use global
+  use material_header,        only: Material
```
This can also be removed

paulromano commented on src/dd_init.F90:
```diff
@@ -0,0 +1,325 @@
+    integer :: neighbor_meshbin
+    integer :: alloc_err        ! allocation error code
+    integer :: world_group
+    integer :: domain_master_group
```
declarations of `world_group` and `domain_master_group` should be surrounded with `#ifdef MPI`

paulromano commented on src/finalize.F90:
```diff
@@ -22,7 +24,7 @@ module finalize
 
   subroutine finalize_run()
-
+    integer :: i
```
`i` is unused

paulromano commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1076 @@
+    integer    :: alloc_err      ! allocation error code
+    integer(8) :: new_source     ! size of the source_bank after distribution
+    integer                 :: current_request  ! current communication request
+    type(VectorInt)         :: requests         ! communication requests
```
`current_request` and `requests` should be surround with `#ifdef MPI`

paulromano commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1076 @@
+    ! need to keep track of how many particles we've sent to each so that we can
+    ! rotate through processes to send to
+    allocate(n_send_domain(n_procs))
+    n_send_domain = 0
```
`n_send_domain(:) = 0` for clarity

paulromano commented on src/dd_comm.F90:
```diff
@@ -0,0 +1,1076 @@
+    ! We're going to send sites one at a time, then ALLREDUCE the number of
+    ! sent sites so each rank knows how many revcs to post
+    allocate(n_send_rank(n_procs))
+    n_send_rank = 0
```
`n_send_rank(:) = 0` for clarity

paulromano commented on src/dd_header.F90:
```diff
@@ -0,0 +1,109 @@
+  type, public :: dd_type
+
+    ! Domain mesh information
+    type(RegularMesh), pointer :: mesh => null()
```
Since there is always a mesh, this should just be a scalar variable.

liangjg replied:
Excuse me, @paulromano , after changing this to `type(RegularMesh) :: mesh`, how to deallocate this `mesh` in `subroutine deallocate_dd(this)`? Thanks.

paulromano commented on src/dd_header.F90:
```diff
@@ -0,0 +1,109 @@
+    if (allocated(this % n_scatters_domain)) &
+         deallocate(this % n_scatters_domain)
+    if (allocated(this % n_scatters_local)) deallocate(this % n_scatters_local)
+    if (allocated(this % scatter_offest)) deallocate(this % scatter_offest)
```
You should also have a call to `this % bins_dict % clear()`

smharper commented on src/resize_arr.F90:
```diff
@@ -0,0 +1,280 @@
+
+      current_size = requested_size
+
+    end if
```
Okay, we can keep `resize_array`, but I think we should
* Remove the `current_size` argument
* Remove the `if` statement so that it can make arrays smaller
* And rename it to `reallocate_array`

smharper commented on src/resize_arr.F90:
```diff
@@ -0,0 +1,280 @@
+
+      current_size = requested_size
+
+    end if
```
Remove `current_size` from `extend_array` and make it call `fatal_error` if `requested_size < size(array)`

liangjg replied:
I found the only difference of `extend_array ` and `resize_array ` is the former keeps the content while the latter doesn't. Maybe it is better to combine them to be one fucntion, with an additional argument `logical keep_data`, like this (yes, the `current_size` will be removed):

```
  subroutine extend_array_bank(array, new_size, keep_data, alloc_err)

    type(Bank), allocatable, intent(inout)     :: array(:)
    integer, intent(in)                        :: new_size
    logical, intent(in)                        :: keep_data
    integer, intent(out)                       :: alloc_err

    type(Bank), allocatable                    :: tmp(:)

    if (new_size < size(array)) then
      call fatal_error("Trying to EXTEND array to a smaller size.")
    end if

    if (new_size > size(array)) then
      allocate(tmp(new_size), STAT=alloc_err)
      if (keep_data) tmp(1:current_size) = array
      call move_alloc(FROM=tmp, TO=array)
    end if

  end subroutine extend_array_bank
```
What do you think? @smharper

smharper replied:
Yes, that looks good.

paulromano commented:
@liangjg Are secondary particles correctly handled now? I seem to remember in December that you said it was not supported at the time. Also, can you elaborate on your initial comment about "turning off random number re-use for calculating URR cross sections"?

liangjg commented:
@paulromano  Currently, the merged code uses `yield * weight` to treat secondary particle. The obstacle is that the particle maybe cross current domain, leaving secondary particles in previous process. If domain decomposition must be implemented in conjunction with multiplicity, the code should be modified to:
  - continue tracking secondary particles after a crossing-domain particle is transferred to another process. 
  - assign a random number seed for every secondary particle (when it is born) to maintain reproducibility.

I can try to implement this later on if you think necessary.

|

Secondly, the change "turning off random number re-use for calculating URR cross sections" is written by Nick. In my understanding it means, the `same_nuc` in `calculate_urr_xs()`, which is used to re-use random number when calculating URR cross sections, will never be true when run in dd mode:

```
    same_nuc = .false.
    do i = 1, nuc % nuc_list % size()
      if (E /= ZERO .and. E == micro_xs(nuc % nuc_list % data(i)) % last_E) then
        same_nuc = .true.
        same_nuc_idx = i
        exit
      end if
    end do

    if (same_nuc) then
      r = micro_xs(nuc % nuc_list % data(same_nuc_idx)) % last_prn
    else
      r = prn()
      micro_xs(i_nuclide) % last_prn = r
    end if

``` 
because the `last_E` is reset to zero evry time when calculating xs:


```
        if (dd_run) then
          ! NOTE: calculate_xs does not re-calculate XS on a per-nuclide
          ! basis if the energy hasn't changed. For DD reproducibility we'd
          ! need to send a prn seed for each nuclide along with particles as
          ! they cross domain boundaries.  This would just about double the
          ! amount of info that goes with particles across boundaries, so we
          ! disable it for DD runs.  If we care about XS calculation speed more
          ! than network communication load, we might consider communicating
          ! these seeds with particles.  Or, we could comment out these two
          ! lines and lose random number reproducibility.
          micro_xs % last_E = ZERO
          micro_xs % last_index_sab = NONE
        end if
        call calculate_xs(p)
```

In fact I think this is the second reason causing dd code cannot get identical results as normal runs (the first reason is the change of random number serial at the beginning of particle transport).

Anyway, this change is simple and will not change anything about the normal runs in non-dd cases.

walshjon commented:
As far as the physics of the URR change, it won't matter for most of the
problems people are currently running, but if there's a detailed gradation
of temperatures across cells, as you might have in a run which requires DD,
you can get some unexpected results.  Not a reason to not go forward with
this, but I think it's worth documenting.

On Mon, Jan 25, 2016 at 2:48 PM, Jingang Liang <notifications@github.com>
wrote:

> @paulromano <https://github.com/paulromano> Currently, the merged code
> uses yield * weight to treat secondary particle. The obstacle is that the
> particle maybe cross current domain, leaving secondary particles in
> previous process. If domain decomposition must be implemented in
> conjunction with multiplicity, the code should be modified to:
>
>    - continue tracking secondary particles after a crossing-domain
>    particle is transferred to another process.
>    - assign a random number seed for every secondary particle (when it is
>    born) to maintain reproducibility.
>
> I can try to implement this later on if you think necessary.
>
> Secondly, the change "turning off random number re-use for calculating URR
> cross sections" is written by Nick. In my understanding it means, the
> same_nuc in calculate_urr_xs(), which is used to re-use random number
> when calculating URR cross sections, will never be true when run in dd mode:
>
>     same_nuc = .false.
>     do i = 1, nuc % nuc_list % size()
>       if (E /= ZERO .and. E == micro_xs(nuc % nuc_list % data(i)) % last_E) then
>         same_nuc = .true.
>         same_nuc_idx = i
>         exit
>       end if
>     end do
>
>     if (same_nuc) then
>       r = micro_xs(nuc % nuc_list % data(same_nuc_idx)) % last_prn
>     else
>       r = prn()
>       micro_xs(i_nuclide) % last_prn = r
>     end if
>
>
> because the last_E is reset to zero evry time when calculating xs:
>
>         if (dd_run) then
>           ! NOTE: calculate_xs does not re-calculate XS on a per-nuclide
>           ! basis if the energy hasn't changed. For DD reproducibility we'd
>           ! need to send a prn seed for each nuclide along with particles as
>           ! they cross domain boundaries.  This would just about double the
>           ! amount of info that goes with particles across boundaries, so we
>           ! disable it for DD runs.  If we care about XS calculation speed more
>           ! than network communication load, we might consider communicating
>           ! these seeds with particles.  Or, we could comment out these two
>           ! lines and lose random number reproducibility.
>           micro_xs % last_E = ZERO
>           micro_xs % last_index_sab = NONE
>         end if
>         call calculate_xs(p)
>
> In fact I think this is the second reason causing dd code cannot get
> identical results as normal runs (the first reason is the change of random
> number serial at the beginning of particle transport).
>
> Anyway, this change is simple and will not change anything about the
> normal runs in non-dd cases.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/559#issuecomment-174637660>.
>



-- 

-------------------------------------------
Jonathan A. Walsh
Ph.D. Candidate
Department of Nuclear Science & Engineering
Massachusetts Institute of Technology
walshjon@mit.edu
(503)505-1872

liangjg commented:
@walshjon Thanks for your comments. My understanding is the random number will not be re-used in dd runs, i.e., new random numbers will be used to calculate URR xs. You mean this may cause wrong results if dealing with nuclides with different temperatures? I found the comments in the code(in `calculate_urr_xs()` ), but I don't understand the `correlation of temperature in probability tables`. Is it caused by breaking this `'correlation' `? Thanks.
```
    ! if we're dealing with a nuclide that we've previously encountered at
    ! this energy but a different temperature, use the original random number to
    ! preserve correlation of temperature in probability tables
    same_nuc = .false.
    do i = 1, nuc % nuc_list % size()
      if (E /= ZERO .and. E == micro_xs(nuc % nuc_list % data(i)) % last_E) then
        same_nuc = .true.
        same_nuc_idx = i
        exit
      end if
    end do
```

liangjg commented:
Hi, @paulromano , I encountered compiling failure again in Travis tests: only in `'phdf5-omp-debug'` mode. It passed on the nsecluster. I deleted another 2 `'save'` but it had no effect. As there is no error message output in Travis, I can't figure out the reason. Could you give me a hint? Thanks.

paulromano commented:
@liangjg Here's a [Travis build with more output](https://travis-ci.org/paulromano/openmc/builds/105069059). It looks like it's still running into a weird internal error compiling the global module. I'll see if I can come up with some way to get around it.

Regarding the URR issue, perhaps I take a more strict viewpoint on this. Any feature that results the physics being wrong (slight as it may be) is not ready to be in the develop branch in my opinion. Otherwise, we will have to say "you can use domain decomposition for large problems, but your answers might not be correct." While it is true that most problems we run today won't have issues, it is exactly the type of problems that you'd want to use domain decomposition on (large reactor with real temperature distribution) that will have problems.

I also feel it is necessary to address the secondary particle issue now rather than merge it as is with a promise to deal with it later. I'm planning on PRing photon transport within the next few months and at that point we can't just rely on the multiplicity trick.

liangjg commented:
Thanks @walshjon , I saw the previous PR #282 .

liangjg commented:
@paulromano I think it can handle secondary particles in the new way now. I just assigned new random number seed for every secondary particle in the same way as in creation of fission neutrons. And the secondary particles are transported from the new random seeds.  I ran BEAVRS in dd case, making sure the eigenvalue is 'right' .

liangjg commented:
Oh sorry, I made a mistake. I reverted the previous commit because it didn't solve tracking of all secondary particles and fixing it seems not straightforward. The current tracking strategy of domain decomposition is somehow based on the idea that only one particle is tracked once. To keep on tracking secondary particles after one particle had crossed the domain, it seems necessary to create some new data structure and change transport flow for dd. I'll think out this later on. Besides, more time is needed to figure out the way to fix urr problem.

liangjg commented:
I fixed the URR problem firstly. As commented in last commit:

> Delete unnecessary reseting micro_xs % last_E and micro_xs % last_index_sab when calculating xs in dd. This change amends the incorrect truning off random number re-use for calculating URR cross sections for the same nuclide at different temperatures. And the efficency is improved by 20%(tranport time) for dd run of beavts benchmark.

As in new dd implementation (I modified before), particles crossing domain are recovered after being transferred to the new domain, and the random number seeds both for transport and xs are communicated and recovered as well, there is no need to re-calculated xs every time if the particle energy hasn't changed, i.e. in the same way as normal runs. Through this amendment, the calculate xs module are totally unchanged compared to normal code. Furthermore, it improves the efficiency(20% for beavrs) because of  eliminating re-calculation of xs.

Thanks for pointing out this problem. @walshjon  @paulromano

liangjg commented:
And it is a surprise that this commit solved the Travis test problem.

smharper commented on src/constants.F90:
```diff
@@ -353,12 +353,45 @@ module constants
+       SRC_ENERGY_MONO    = 1, & ! Monoenergetic source
+       SRC_ENERGY_MAXWELL = 2, & ! Maxwell fission spectrum
+       SRC_ENERGY_WATT    = 3, & ! Watt fission spectrum
+       SRC_ENERGY_TABULAR = 4    ! Tabular distribution
```
Remove these source parameter variables

smharper commented on src/finalize.F90:
```diff
@@ -1,8 +1,10 @@
 module finalize
 
+  use constants
```
Remove the `use constants`

smharper commented on src/output.F90:
```diff
@@ -911,6 +913,33 @@ subroutine print_overlap_check
+    ! display header block for geometry debugging section
+    if (master) then
+      call header("Domain Interaction Counts")
+      write(ou,100) 'Domain Mesh I-J-K','No. Interactions'
```
We tend to avoid labels like this.  Change this line to
```f90
write(ou, "(1X,A,T30,A)") 'Domain Mesh I-J-K','No. Interactions'
```

smharper commented on src/output.F90:
```diff
@@ -911,6 +913,33 @@ subroutine print_overlap_check
+    ijk = trim(to_str(domain_decomp % ijk(1))) // "-" // &
+          trim(to_str(domain_decomp % ijk(2))) // "-" // &
+          trim(to_str(domain_decomp % ijk(3)))
+    write(ou,101) trim(ijk), domain_decomp % interaction_count
```
Similar treatment for this line

smharper commented on src/output.F90:
```diff
@@ -911,6 +913,33 @@ subroutine print_overlap_check
+    write(ou,101) trim(ijk), domain_decomp % interaction_count
+
+100 format (1X,A,T30,A)
+101 format (1X,A,T30,I12)
```
Then delete these lines

smharper commented on src/source.F90:
```diff
@@ -45,6 +47,11 @@ subroutine initialize_source()
 
+      if (dd_run) then
+        call fatal_error("Reading source from binary file not " // &
+                         "implemented for domain decomposition.")
```
The `//` isn't needed to concatenate strings
```f90
        call fatal_error("Reading source from binary file not &
                          &implemented for domain decomposition.") 
```

smharper commented:
Rename
test_domain_decomp.py ===> test_dd.py
test_domain_decomp_otf_tallies.py ===> test_dd_otf_tallies.py

smharper commented on src/dd_header.F90:
```diff
@@ -0,0 +1,110 @@
+    logical    :: count_interactions = .false.
+    integer(8) :: interaction_count = 0_8
+
+  end type DomainDecomType
```
May as well make `deallocate_dd` a type-bound procedure
```f90
    contains
    procedure(deallocate_dd) :: deallocate
```

smharper replied:
Similarly,
```f90
    procedure(initialize_domain_decomp) :: initialize
```

liangjg replied:
@smharper If `initialize_domain_decomp` is made a type-bound procedure, `'use global'` is needed in `module dd_header`, but `'use dd_header'` exists in `module dd_header`, causing compiling failure `'Circular dependency'`. What's your suggestion to this?

smharper replied:
I guess it's not worth the trouble.  Just leave `initialize_domain_decomp` as a non-type-bound procedure

smharper commented on src/dd_init.F90:
```diff
@@ -0,0 +1,327 @@
+            mesh_indices_to_bin(dd % mesh, dd % ijk + (/ 0,  1,  1/)), & !+y
+            dd % meshbin                                             , & !-z
+            mesh_indices_to_bin(dd % mesh, dd % ijk + (/ 0,  0,  2/))  & !+z
+                 /)
```
Could probably make this more short, clear, and less error-prone with a set of `do` loops.

liangjg replied:
`do` loops have been used to replace these lines.
```
    integer, dimension(3, 7) :: indices ! Neighbor indices differences
    integer                  :: i,j     ! Loop index

    indices(1:3, 1) = (/ 0,  0,  0/) ! self, for first order
    indices(1:3, 2) = (/-1,  0,  0/) ! -x
    indices(1:3, 3) = (/ 1,  0,  0/) ! +x
    indices(1:3, 4) = (/ 0, -1,  0/) ! -y
    indices(1:3, 5) = (/ 0,  1,  0/) ! +y
    indices(1:3, 6) = (/ 0,  0, -1/) ! -z
    indices(1:3, 7) = (/ 0,  0,  1/) ! +z

    ! Loop to set first-order (i = 1) and second-order (i = 2-7) neighbors
    do i = 1, 7
      do j = 1, 6
        dd % neighbor_meshbins( (i - 1) * 6 + j ) = mesh_indices_to_bin( &
             dd % mesh, dd % ijk + indices(1:3, i) + indices(1:3, j + 1))
      end do
    end do
```

smharper commented on src/output.F90:
```diff
@@ -877,6 +879,33 @@ subroutine print_overlap_check
+    write(ou, "(1X,A,T30,I12)") trim(ijk), domain_decomp % interaction_count
+
+100 format (1X,A,T30,A)
+101 format (1X,A,T30,I12)
```
Delete these format lines

smharper commented:
@paulromano, it would really make DD simpler if the secondary particles were individual particles that each have `transport` called on them.  And that could be advantageous simply from the conceptual perspective of `transport` operating on one particle at a time.  We could handle that if we make the `secondary_bank` a global variable that we loop over in simulation.F90.  Of course that would mean allocating the memory for the secondary bank.  Thoughts?

paulromano commented:
@smharper I recall having a similar conversation in #445. There are a few problems. First is that if secondary particles become part of the main loop, the loop size is then non-deterministic which will making dividing work among threads very difficult. The other issue is that if you have a single global secondary bank (per process), you'll have multiple threads writing to it simultaneously, so you'd have to protect writes with a critical section which may have a detrimental impact on thread scalability.

I still find the idea of `transport` being the simulation of a single source particle and any secondaries it creates the cleanest, so at this point in time I'm not inclined to change that structure.

smharper commented:
@paulromano, ah so we did.  Fair enough.  We will find another way, then.

liangjg commented:
The last commit [c4f35afe](https://github.com/liangjg/openmc/commit/c4f35afe6107f2b2f9eb386691e55e2f28daf900) is to support secondary particles in dd runs. It is implemented via:
  + assigning one random seed for every secondary particle for reproducibility;
  + keeping on tracking secondary particles after older particle died or even crosses domain.

It should be noted that secondary particles are transported after mother particle's death, using the same particle variable `p` as the mother particle. But in dd runs the crossing domain information are stored in this particle variable too, which means the dd data would be covered when transporting secondary particles. To avoid this, particles are grabbed out from particle buffer for tracking (not using the address) and every particle would be stored back to particle_buffer when it flies out of domain. The particle buffer should be extended when necessary.

@paulromano  @smharper

liangjg commented:
However I have to bring back the `"urr"` issue. Tests on simple models showed the latest version can get 'right' and identical results with the `re-using` of urr random number (and with secondary particles). But when it comes to BEAVRS, the reproducibility was broken. By tracking the random number sequence I found it was related to `calculate_xs`. It seems to me that reproducibility and re-using URR random number is difficult to coexist. I try to explain the situation here:

`micro_xs` is the cache to store micro xs for all nuclides. Last neutron energies `last_E ` (for all nuclides) are also stored to avoid re-calculation when the energy hasn't changed. But in dd runs, a particle is transferred to another process when it crosses domain, while these `micro_xs` are not transferred (it is apparently not appropriate to communicate all the `micro_xs` together with a particle ). As the cache 'micro_xs' is probably different in the new domain, more or less xs calculations will be made for transferred particles. For example, after a repeated crossing surface event with the energy unchanged,  `calculate_xs` for the some nuclides are skipped but it is not the case if the particle cross to another domain. And if some nuclides use p-table in `urr`, the random number sequence would change.

In original dd version, we force calculating xs by reseting  `last_E` to be zero before every `event`. But it turns off random number re-using for correlation in `urr`. In the commit [56493c22](https://github.com/liangjg/openmc/commit/56493c22845f160900da28a03cda7b5ab1bd7472), I deleted the forcing treatment as I thought the `micro_xs` are re-calculated before a crossing-domain particle being tracked in the new domain. However it only applies to simple cases. In other cases which involve a lot of nuclides, there is always possibility that the judgement of `E` and `last_E` goes wrong.

Now I can only see two choices: one is the original strategy to keep reproducibility, bearing more xs calculations and inability of `urr correlation`; the other is losing random number reproducibility. Maybe you can think of a better way. I'm expecting your suggestions. Thanks. @paulromano  @smharper

smharper commented:
@liangjg, I want to propose a change to the DD algorithm.  I have the following concerns which prompted this thought:
* `Particle` is a really big object so it is probably quite inefficient to communicate.
* DD adds a lot of bloat to some of our most basic code like the `Particle` object and the `transport` subroutine.
* The logic for translating particles when they cross domains is somewhat convoluted.  I think it will be difficult to keep that code working as we add new features.
* Similarly, the prng stream is a little harder to understand for DD runs.  Again, this makes it difficult for developers to keep everything working as we make changes.

What I propose is this: when a particle leaves a domain, simply translate it to that domain boundary and communicate a `Bank` rather than a `Particle`.  Unfortunately, this means that changing the number of domains will change the results.  But I think that irreproducibility is worth the benefits.  I believe it will mitigate the concerns I listed above.  Thoughts?

liangjg commented:
@smharper , thanks for your reply. 
Firstly, I want to note that `Particle` is not used as the communication object. A new type `ParticleBuffer` is created for particle communication. It is a part of `Particle` therefore much thinner (about 200 bytes), though larger than `Bank` (76 bytes). As you are concerned, I also doubt wether all the attributes in `ParticleBuffer` are necessary for communication, but I don't think it appropriate to use `Bank` to replace `ParticleBuffer`, because at least some of the attributes are useful. For example, the attribute `cell_born` is necessary for tally `Birth Cell`. We cannot extend `Bank` if want to make dd support all features. I can try to verify all the other attributes and simplify `ParticleBuffer` to improve communication efficiency as much as possible, even though the communication efficiency is evaluated acceptable in Nicholas's thesis.

Secondly, `when a particle leaves a domain, simply translate it to that domain boundary` is the original way in dd implementation, which definitely cannot promise reproducibility. I did make some changes to get reproducibility, including:
- When a particle is leaving a domain, (do not move the particle to a domain boundary) transfer it to new domain and recover it for continuing transport, with tracking distance transferred together with it to avoid sampling again
- Account for situation when a particle crosses beyond domain neighbors (crossing 2 or more domains one time). Communicate the particle again when it is not in current domain
- `calibrate` coordinates whenever a particle move forward, to make sure particle coordinates of different level consistent with each other, and thereby particle is recovered exactly
- Fix locating differently when a particle is on a surface, by moving slightly forward to get the cell info and recovering coordinates afterwards
I agree the changes somewhat complicate the codes, mainly the `transport` process. I don't mind to move back if we determine the reproducibility is less important. The concern I can now think of is we may have to make the dd tests run in fixed process/threads to get identical results.

But I have a request here, could you please write all your concerns/suggestions about this dd merge before I make more changes? I myself think Nicholas had done great work and I'm really desired to get it merged but if dd is considered quite not mature to merge into openmc, maybe we'd better find another way (doing some more research and rewrite it?).

Tomorrow is the Chinese Spring Festival. Best wishes to all of you in the year of the monkey. Thanks.

smharper commented:
@liangjg Sorry for the delay in this reply.

My biggest concern is really the the readability/maintainability of our core subprograms.  OpenMC needs to be easy to understand and modify by students who aren't going to be around for a long time.

In this particular case, I just don't think the reproducibility across different numbers of domains is worth all the added code e.g. to `transport`.  I'd like to hear if other developers agree with that sentiment.  Paging @paulromano.  Sorry, I know you probably spent a lot of time trying to achieve that reproducibility.

Regarding your request for me to write all of my concerns, here are the major ones I haven't voiced so far:
* How will restart work with multiple statepoint files?
* I think OpenMC should concatenate all of the statepoint files at the end of the run, or at a bare minimum, there should be a script that can concatenate them in post-processing.
* The PythonAPI needs to be modified so it can read the results from OTF tallies.
* The PythonAPI also needs to be modified so it can make the DD-related XML inputs.

wbinventor commented:
For what it's worth, I'll chime in with my two cents about reproducibility. Although I haven't reviewed this PR, I would like to see a move made on the major PRs like this one that are waiting in limbo.

I personally believe that the loss of reproducibility would make it more difficult for users to debug their inputs / models in the future. As I am sure is the case for many other code users out there, I always find it best in general to confirm that small models / simulations matched expected results before scaling up the number of histories and processes by a factor of 1000 or more. In the case of DD, this *might* look like the following (fully enumerated but perhaps simplified) sequence of simulations:

1) Run small sequential simulation without DD on local machine
2) Run small parallel simulation without DD on local machine
3) Run small parallel simulation with DD on local machine
4) Run small parallel simulation without DD on cluster
5) Run small parallel simulation with DD on cluster
6) Run HUGE parallel simulation with DD on cluster

If everything was reproducible, the results from stages 1-5 should all be identical, providing the user with some confidence in their input model before pressing "go" on stage 6. 

From my perspective, the vast majority of the user's time is spent iterating between 1-5 to debug an input model. The vast majority of the compute time is spent at 6 at which point the reproducibility becomes less of an issues since one may not have multiple "hero runs" to compare between. Since I think a human's time is more valuable than that of a computer, my personal take is that we should determine the value of reproducibility to a human user in stages 1-5. 

From my experience, it is highly important to retain reproducibility in these stages. The tally uncertainties in these debugging stages are huge, so if any discrepancies arise between iterations of stages 1-5, it will unclear if they arose due to perturbations the user made to the input files or rather due to the lack of reproducibility inherent to the simulation itself. A user's confidence in the code will go down if they must always run "hero" calculations in order to see the changes made to their inputs "emerge" as a true bias in their tallies between inputs (rather than statistical fluctuations in OpenMC).

I know this argument takes more of a "user-centric" approach to the issue and ignores @smharper's very good points about code simplicity and modularization. But without further debate, my vote would be to retain all of the hard work that @liangjg and @nhorelik have put into maintaining reproducibility with DD.

wbinventor commented:
As a follow up point to this, the loss of reproducibility may make things even more challenging to those of us using OpenMC as an upstream "nuclear data" code to generate inputs into downstream deterministic codes (*i.e.*, MGXS generation for OpenMC / Rattlesnake / MOCingjay, reaction rates for depletion, etc.). In my mind it is essential that we can have confidence that changes in the results from those codes which operate downstream of OpenMC are due to our changes made in those downstream codes themselves *rather* than irreproducible error in OpenMC (talk about debug hell). Again, this is a non-issue if one only runs "hero" calculations which  "wash" out the irreproducible statistical fluctuations in the tallies. But that may not (in my experience is not) be what most users want to do during the debug phase of their projects.

siegela67 commented:
I basically agree. CS people will argue though that bitwise reproducibility is setting the bar too 
unrealistically high anyhow on next generation architectures, operating much closer to threshold
voltage and with huge power cost of ECC, some non-determinism will be inherent, and that
you need to develop debugging strategies based on statistical validation metrics, and that bitwise
is unrealistic and will be a thing of the past. I’ll believe it when I see it, though, am highly skeptical
that chip designers will produce such garbage. Maybe as an option (turn off ECC in exchange for more cores)
but it’s something to at least keep in mind. Otherwise, in every large code project I’ve been part of we’ve
retained a bitwise reproducibility option for debugging that was typically turned off for production runs 
(large or otherwise, size doesn’t really matter). this was often hand rolled implementations of MPI collectives
based on deterministic point to point sequences among other things. I found it incredibly valuable. In fact,
we had a tool once, called Alice, that ran two simultaneous images of code, connected their memory, and 
printed line where first bit of divergence occurred in specified data structure. Saved me so many times.

So I agree with what you say, but I’d explore the argument in terms of what Intel and NVIDIA have in mind on the 5-10
year time frame.

I’m also not convinced DD has any use cases relative to other options, but that’s another point.
On Feb 20, 2016, at 8:53 AM, Will Boyd <notifications@github.com> wrote:

> For what it's worth, I'll chime in with my two cents about reproducibility. Although I haven't reviewed this PR, I would like to see a move made on the major PRs like this one that are waiting in limbo.
> 
> I personally believe that the loss of reproducibility would make it more difficult for users to debug their inputs / models in the future. As I am sure is the case for many other code users out there, I always find it best in general to confirm that small models / simulations matched expected results before scaling up the number of histories and processes by a factor of 1000 or more. In the case of DD, this might look like the following (fully enumerated but perhaps simplified) sequence of simulations:
> 
> 1) Run small sequential simulation without DD on local machine
> 2) Run small parallel simulation without DD on local machine
> 3) Run small parallel simulation with DD on local machine
> 4) Run small parallel simulation without DD on cluster
> 5) Run small parallel simulation with DD on cluster
> 6) Run HUGE parallel simulation with DD on cluster
> 
> If everything was reproducible, the results from stages 1-5 should all be identical, providing the user with some confidence in their input model before pressing "go" on stage 6.
> 
> From my perspective, the vast majority of the user's time is spent iterating between 1-5 to debug an input model. The vast majority of the compute time is spent at 6 at which point the reproducibility becomes less of an issues since one may not have multiple "hero runs" to compare between. Since I think a human's time is more valuable than that of a computer, my personal take is that we should determine the value of reproducibility to a human user in stages 1-5.
> 
> From my experience, it is highly important to retain reproducibility in these stages. The tally uncertainties in these debugging stages are huge, so if any discrepancies arise between iterations of stages 1-5, it will unclear if they arose due to perturbations the user made to the input files or rather due to the lack of reproducibility inherent to the simulation itself. A user's confidence in the code will go down if they must always run "hero" calculations in order to see the changes made to their inputs "emerge" as a true bias in their tallies between inputs (rather than statistical fluctuations in OpenMC).
> 
> I know this argument takes more of a "user-centric" approach to the issue and ignores @smharper's very good points about code simplicity and modularization. But without further debate, my vote would be to retain all of the hard work that @liangjg and @nhorelik have put into maintaining reproducibility with DD.
> 
> —
> Reply to this email directly or view it on GitHub.
>

smharper commented:
I am swayed by @wbinventor's and @siegela67's views.  But we definitely need to clean up `transport`.  I'll try to provide suggestions for clarity and readability improvements.

smharper commented on src/tracking.F90:
```diff
@@ -85,20 +119,121 @@ subroutine transport(p)
+      end if
+
+      ! Check for domain boundary crossing
+      if (dd_run .and. d_dd_mesh <= distance) then
```
Currently `transport` looks something like this:
```
Check for a domain crossing (exit if so)
Get distance to boundary and collision
Check for a domain crossing (exit if so)
If surface crossing:
  Handle surface crossing
Else (it is a collision):
  Handle collision
Repeat
```

I would prefer it to look more like this
```
Get distance to boundary, collision, and domain crossing
If surface crossing:
  Handle surface crossing
Else if collision:
  Handle collision
Else (it is a domain crossing):
  Handle domain crossing
  (kill the particle rather than exit the loop)
Repeat
```

smharper replied:
Maybe we should stop using `if` statements for that block of code and use a `select case`.  Declare some constant parameters like
```f90
integer, parameter :: PARTICLE_CROSSES_SURFACE=1
integer, parameter :: PARTICLE_COLLIDES=2
integer, parameter :: PARTICLE_CROSSES_DOMAIN=3
```
We could determine which is the case by
```f90
if (dd_run) then
  if (p % stored_distance > ZERO)
    xyz_temp = p % coord(1) % xyz
    p % coord(1) % xyz = xyz_temp + p % fly_dd_distance * p % coord(1) % uvw
    call distance_to_mesh_surface(p, domain_decomp % mesh, &
         distance, d_dd_mesh, meshbin=domain_decomp % meshbin)
    p % coord(1) % xyz = xyz_temp
  else
    call distance_to_mesh_surface(p, domain_decomp % mesh, &
         distance, d_dd_mesh, meshbin=domain_decomp % meshbin)
end if

if (dd_run .and. d_dd_mesh < d_boundary .and. d_dd_mesh < d_collision) then
  event = PARTICLE_CROSSES_DOMAIN
else if (d_boundary < d_collision) then
  event = PARTICLE_CROSSES_SURFACE
else
  event = PARTICLE_COLLIDES
end if
```
And then use a select case
```f90
select case(event)
case (PARTICLE_CROSSES_SURFACE)
  ...
case (PARTICLE_COLLIDES)
  ...
case (PARTICLE_CROSSES_DOMAIN)
 ...
end select
```

smharper commented on src/tracking.F90:
```diff
@@ -85,20 +119,121 @@ subroutine transport(p)
+        p % coord(1) % xyz = xyz_temp + p % fly_dd_distance * p % coord(1) % uvw
+        call distance_to_mesh_surface(p, domain_decomp % mesh, &
+             distance, d_dd_mesh, meshbin=domain_decomp % meshbin)
+        p % coord(1) % xyz = xyz_temp
```
Rather than modifying `p % coord(1) % xyz` here in `transport`, I think we should just use a shifted `xyz` inside `distance_to_mesh_surface` itself.  Am I correct in thinking that `p % stored_distance` is exactly zero when we make the other call to `distance_to_mesh_surface`?  If so, then we can check `p % stored_distance` inside `distance_to_mesh_surface` to see if the coordinates need to be offset or not.

liangjg commented:
@smharper Thanks for your very good suggestions to improve clarity and readability. The dd-related treatments in the `transport` subroutine have been cleaned up as you suggested, except a logical variable `dd_boundary_crossed` instead of 'select' is used to judge whether it is a normal event (`collision` or `crossing surface`) or `crossing domain` event.
And, the new LCG URR approach as well as multi-group feature has been merged  into dd version.
Looking forward to more suggestions.

smharper commented on src/particle_header.F90:
```diff
@@ -141,6 +186,60 @@ subroutine initialize_particle(this)
+    type(ParticleBuffer), intent(in)  :: buf
+    type(Particle),       intent(out) :: part
+
+    part % id              = buf % id
```
Add a `call part % initialize()` at the start of this routine so we can be sure that there is no left-over state from the last particle.

smharper commented on src/tracking.F90:
```diff
@@ -64,6 +74,18 @@ subroutine transport(p)
+        call fatal_error("Could not locate particle after crossing domain.")
+      end if
+      p % coord(1) % xyz = xyz_temp
+    end if
```
I think we can delete this code block.  `find_cell` takes into account the particle's direction so it should properly handle surfaces without the need to shift by `TINY_BIT`.  I changed this block to say
```f90
    if (dd_run .and. current_stage > 1) then
      p % coord(p % n_coord) % cell = NONE
    end if
```
(so we would call `find_cell` inside the event loop) and the tests pass.  Even the above code should be unnecessary if you follow my suggestion of initializing particles before loading from buffer.

liangjg replied:
Yeah, I tried to remove these lines and tested BEAVRS as I remembered the inconsistent re-locating occurs only at complex geometry when particle crosses domain after just moving to a surface (the coordinate is slightly inside the former cell). But it didn't appear this time (I scaled the population up to 100000). So I think you are right. The `find_cell` was updated to account for particle's direction so re-locating can be correctly handled inside the event loop. @smharper

liangjg commented:
Several improvements were made to the domain decomposition (DD) branch in the last commits:

- [X] Made DD compatible with CMFD acceleration (keep reproducibility regardless of how to divide domains)
- [X] Implemented source bank writing to hdf5 file for DD runs
- [X] Implemented domain meshline plot on top of the geometry
- [X] Updated Python API to generate DD-related XML inputs
- [X] Updated Python API to read results of OTF tallies which are written into domain-separated statepoint files
- [X] A Jupyter Notebook example "Domain Decomposition" was created, introducing the whole process of creating the model, plotting domain mesh and extracting tallies or sources for a DD run.

These updates cover several concerns of @smharper  in his last comment, except additional two on which I think we need more discussions:

1. the concern of restart from multiple statepoint files. Theoretically, restart can be implemented by reading all domain-related information, such as domain partition, domain-specified source sites, on-the-fly tallies and accounting for the possible change of parallel processes. But I'd like to listen to more comments about the necessity.
2. the complexity caused by achieving reproducibility. Thanks @wbinventor  and @siegela67  for your comments. I myself find reproducibility is vitally important for DD implementation, especially for debugging. Regarding the complexity, DD itself is closely related to the particle transport so it inevitably affects "core subprograms" of openmc, such as `transport`. Complexity would remain even if reproducibility is abandoned.

Looking forward to your suggestions/comments. @smharper @paulromano

liangjg commented:
Some more updates were made in this dd branch:
- clearing up statepoint files: the domain-separated statepoint files is deprecated.  All domain-specified, on-the-fly tallies are written into the single statepoint
- supporting restart calculation for dd runs: the code writes all dd-related data and continues the simulation from a statepoint while keeping reproducibility
- other improvements/refactoring

liangjg commented:
Well, I just ran the BEAVRS to test the latest domain decomposition features. Here I'd like to post the results, as a summary for this long-term PR.

Modeling
==========

Four test cases were run to make the comparisons.

| case        | ref       | dd_off    | dd_1d     | dd_12d    |
|-------------|-----------|-----------|-----------|-----------|
| code        | develop   | dd_branch | dd_branch | dd_branch |
| dd ative    | None      | No        | Yes       | Yes       |
| domain mesh | None      | None      | 1x1x1     | 2x2x3     |

In the modeling, the only change of the inputs is adding a `domain_decomposiiton` element in the `settings.xml` file.

``` xml
    <domain_decomposition>
        <mesh>
            <dimension>2 2 3</dimension>
            <lower_left>-201.63 -201.63 0.0</lower_left>
            <upper_right>201.63 201.63 460.0</upper_right>
        </mesh>
        <allow_leakage>true</allow_leakage>
        <count_interactions>true</count_interactions>
    </domain_decomposition>
```

Domain decomposition for dd_12d case is shown here:
![radial_mats_midplane](https://cloud.githubusercontent.com/assets/3030455/18976596/0c54ce18-8681-11e6-95f9-50c226915e3d.png)
![axial_mats_row_8](https://cloud.githubusercontent.com/assets/3030455/18976599/104975fa-8681-11e6-9a82-ef8199d5277b.png)

- Eigenvalue: 1E+05 particles/batch x 600 active batches (450 inactive batches) 
- Tallies: tallying flux in all fuel pins 
- Processes: 24


Results
=========

keff
----

| case                      | ref                  | dd_off               | dd_1d                | dd_12d               |
|---------------------------|----------------------|----------------------|----------------------|----------------------|
| k-effective (Collision)   | 1.01101 +/-  0.00014 | 1.01101 +/-  0.00014 | 1.01108 +/-  0.00014 | 1.01108 +/-  0.00014 |
| k-effective (Track-length)| 1.01112 +/-  0.00016 | 1.01112 +/-  0.00016 | 1.01134 +/-  0.00016 | 1.01134 +/-  0.00016 |
| k-effective (Absorption)  | 1.01104 +/-  0.00013 | 1.01104 +/-  0.00013 | 1.01096 +/-  0.00013 | 1.01096 +/-  0.00013 |
| Combined k-effective      | 1.01106 +/-  0.00011 | 1.01106 +/-  0.00011 | 1.01107 +/-  0.00011 | 1.01107 +/-  0.00011 |
| Leakage Fraction          | 0.00003 +/-  0.00000 | 0.00003 +/-  0.00000 | 0.00002 +/-  0.00000 | 0.00002 +/-  0.00000 |


tally
-----

![tally](https://cloud.githubusercontent.com/assets/3030455/18976602/18d3d7ce-8681-11e6-9608-a6798c3f0d82.PNG)



timing
-------

| case                              | ref        | dd_off     | dd_1d      | dd_12d     |
|-----------------------------------|------------|------------|------------|------------|
| Total time for initialization     | 3.9640E+01 | 3.8856E+01 | 3.2827E+01 | 3.0417E+01 |
|   Reading cross sections          | 3.1095E+01 | 3.1678E+01 | 2.7660E+01 | 2.5557E+01 |
| Total time in simulation          | 1.9544E+04 | 1.9386E+04 | 2.0578E+04 | 3.1361E+04 |
|   Time in transport only          | 1.7024E+04 | 1.5924E+04 | 1.7829E+04 | 1.3346E+04 |
|   Time in inactive batches        | 7.0747E+03 | 7.0745E+03 | 7.3895E+03 | 1.0717E+04 |
|   Time in active batches          | 1.2469E+04 | 1.2312E+04 | 1.3189E+04 | 2.0644E+04 |
|   Time synchronizing fission bank | 8.4651E+02 | 1.8106E+03 | 1.5810E+00 | 1.3040E+00 |
|     Sampling source sites         | 6.1800E-01 | 6.8400E-01 | 7.8000E-01 | 2.5700E-01 |
|     SEND/RECV source sites        | 1.0100E-01 | 1.2600E-01 | 4.8200E-01 | 5.6400E-01 |
|   Time accumulating tallies       | 1.6610E+03 | 1.6367E+03 | 1.9503E+03 | 1.1703E+02 |
| Total time for finalization       | 4.9200E-01 | 4.8900E-01 | 5.2900E-01 | 1.2500E-01 |
| Total time elapsed                | 1.9584E+04 | 1.9426E+04 | 2.0612E+04 | 3.1392E+04 |
| Calculation Rate (inactive)       | 6360.67    | 6360.83    | 6089.72    | 4198.91    |
| Calculation Rate (active)         | 4811.99    | 4873.45    | 4549.35    | 2906.36    |


Summary
========
- for non-dd runs, the merge of dd **does not affect the calculation results and effeciency** of the code, seeing `dd_off` v.s. `ref`
- for different domain decomposition schemes of a specific model, **the simulations are reproducible** and the effeciency penalties are acceptable, seeing `dd_1d`  v.s. `dd_12d`
- for dd runs, the results are different from non-dd runs, but they are correct, seeing `dd_1d`/`dd_2d` v.s. `dd_off`/`ref`
- usage of domain decomposition and post-processing of dd runs are easy and simple


Regarding this PR, I think all the comments have been addressed so I'll leave it here, untill more comments are given.

paulromano commented:
This PR was discussed at our developer's meeting last summer. While there is certainly a desire and use case to have this functionality, it is a pretty intrusive change in the source and would require the development team to maintain the capability which we were not ready to do. In the future, it is certainly worth revisiting this to see if there's a way to implement domain decomposition in a way that doesn't require changes throughout the codebase. At this point, this PR is obviously way out of sync, especially with most of the code moving to C++, so I'm going to go ahead and close it. Thanks to @nhorelik and @liangjg for the hard work on this, and I'm really sorry that we couldn't get it in as a main feature of OpenMC.
